### DISCUSSION TITLE
Multiple command objects for the same file

### DISCUSSION THREAD

**Discussion by abdelmaged** [clangd/clangd#2004] [Q&A] [answered]
I'm using clangd in vscode, and I'm working on a project where it's very common to compile the same file for different executables with different defines, i.e. if I have two executables `exe1` & `exe2`, and one file `file.c`, then `exe1` will compile `file.c` with `-Dexe1` and `exe2` will compile `file.c` with `-Dexe2`.

And `file.c` contents is something like:
```
int main() {
 // common code for exe1 & exe2 & ...

#ifdef exe1
  // something specific for exe1
#elif exe2
  // something specific for exe2
#else
  // something specific other executables
#endif

 // common code for exe1 & exe2 & ....

  return 0;
}
```

This produces compile_commands.json like:
```
[
  { "directory": "/home/user/llvm/build",
    "command": "gcc -c -o one/file.o -Dexe1 file.c",
    "file": "file.c" },

  { "directory": "/home/user/llvm/build",
    "command": "gcc -c -o two/file.o -Dexe2 file.c",
    "file": "file.c" },

  ...
]
```
We have one file `file.c` and multiple command object for the same file.

1. How does clangd handle this situation ?
2. How to tell clangd to dynamically pick specific command object for `file.c`, like pick the command with -Dexe2 instead of -Dexe1 ?
3. Is it possible that clangd can merge all command objects into single view ?

**Comment by HighCommander4** [comment, marked as answer]
>    1. How does clangd handle this situation ?

It will use an unspecified one of the available commands for the file (likely the first one it encounters).

>    2. How to tell clangd to dynamically pick specific command object for `file.c`, like pick the command with -Dexe2 instead of -Dexe1 ?

There isn't currently a way to do this.

One thing I have seen people do is perform a post-processing step on the `compile_commands.json` to filter out some entries. [jq](https://github.com/jqlang/jq) can be a helpful tool for this job.

>    3. Is it possible that clangd can merge all command objects into single view ?

It's not clear what merging would mean here.

I can imagine merging the _commands_, such that you get a command containing both `-Dexe1` and `-Dexe2`. However, that can easily lead to compiler errors. For example:

```c++
#ifdef exe1
  int var = 42;
#elif exe2
  int var = 43;  // same variable name
#endif
```

If this is compiled with both `-Dexe1` and `-Dexe2`, we would get a diagnostic like "mutiple declarations of variable `var`".

**Reply by i-ky** [reply]
I would like to add that regarding
>Is it possible that clangd can merge all command objects into single view ?

...there is #681.
